### PR TITLE
Upgrade to 3.x stable

### PR DESCRIPTION
**Pull request by vinniefranco**
### What's new?
* Spree 3.0 support!
* Launchy support for specs

### CODE REVIEW

**Comment by vinniefranco** [comment]
#124

**Comment by yf6** [comment]
@vinniefranco this might be on your to-do list already, but would you mind merging master into 2-4-stable before you merge this PR, so 2-4-stable can have the most recent bug fixes and features that's still 2-4 compatible. Thanks.

**Comment by vinniefranco** [comment]
@yf6 Of course! On it.

**Comment by vinniefranco** [comment]
@yf6 done!

File: app/views/spree/admin/orders/_assemblies.html.erb [hunk 1]
```diff
@@ -1,62 +1,54 @@
...
-      <%= Spree.t(:product_bundles) %>
-    </legend>
-  </fieldset>
+  <div class="panel panel-default">
+    <div class="panel-heading">
+      <h1 class="panel-title">
+        <%= Spree.t(:product_bundles) %>
+      </h1>
+    </div>
+    <table class="table table-bordered product-bundles" data-hook="">
```
Comment (elle) [inline app/views/spree/admin/orders/_assemblies.html.erb]: Looking at the original `table` markup, it only uses `product-bundles` class.
Are the other two classes (`table table-bordered`) really needed?

Comment (vinniefranco) [inline app/views/spree/admin/orders/_assemblies.html.erb]: 3.0 is a switch to bootstrap as their CSS framework. So yes, this follows all other tables in 3.x

File: app/views/spree/admin/orders/_assemblies.html.erb [hunk 2]
```diff
@@ -1,62 +1,54 @@
...
-      <col style="width: 15%;" />
-      <col style="width: 15%;" />
-      <col style="width: 15%;" />
-    </colgroup>
+      <% order.line_items.assemblies.each do |item| %>
+        <tbody>
+          <tr class="" data-item-quantity="<%= item.quantity %>">
+            <td class="item-image"><%= mini_image(item.variant) %></td>
+            <td class="item-name">
+              <%= item.product.name %> - <%= item.variant.sku %> <br />
```
Comment (elle) [inline app/views/spree/admin/orders/_assemblies.html.erb]: What do you think about delegating these two methods calls to `product` and `variant`?
i.e. 
`delegate :name, to: :product, prefix: true` => `<%= item.product_name %>`
`delegate :sku, to: :variant, prefix: true` => `<%= item.variant_sku %>`

File: app/views/spree/admin/orders/_assemblies.html.erb [hunk 3]
```diff
@@ -1,62 +1,54 @@
...
+              <%= "(" + variant_options(item.variant) + ")" unless item.variant.option_values.empty? %>
```
Comment (elle) [inline app/views/spree/admin/orders/_assemblies.html.erb]: Same here (i.e. delegate)

Also, what do think about changing the double negatives to `if... any?`

File: Rakefile [hunk 4]
```diff
@@ -6,7 +6,7 @@ require 'spree/testing_support/common_rake'
 
 RSpec::Core::RakeTask.new
 
-task :default => [:spec]
+task default: [:spec]
```
Comment (elle) [inline Rakefile]: Looking at the views, there are many places where the hash rocket syntax is used. So should we change it here?

File: app/views/spree/admin/orders/_assemblies.html.erb [hunk 5]
```diff
@@ -1,62 +1,54 @@
...
 
-    <thead>
-      <th colspan="2"><%= Spree.t(:item_description) %></th>
-      <th><%= Spree.t(:price) %></th>
-      <th><%= Spree.t(:quantity) %></th>
-      <th><%= Spree.t(:total) %></th>
-      <th class="orders-actions actions" data-hook="admin_order_form_line_items_header_actions"></th>
-    </thead>
+            <td class="line-item-qty-edit is-hidden">
+              <%= number_field_tag :quantity, item.quantity, :min => 0, class: "form-control line_item_quantity", :size => 5 %>
```
Comment (dfreerksen) [inline app/views/spree/admin/orders/_assemblies.html.erb]: Mixed hash syntax. I'm not sure which one Spree prefers.

File: app/views/spree/admin/orders/_stock_item.html.erb [hunk 6]
```diff
@@ -1,45 +1,45 @@
...
+
     <td class="item-name">
       <%= item.variant.product.name %><br><%= "(" + variant_options(item.variant) + ")" unless item.variant.option_values.empty? %>
+      <% if item.variant.sku.present? %>
+        <strong><%= Spree.t(:sku) %>:</strong> <%= item.variant.sku %>
+      <% end %>
+
       <% if item.part && item.line_item %>
-        <i><%= Spree.t(:part_of_bundle, sku: item.product.sku) %></i>
+        <strong><%= Spree.t(:part_of_bundle, sku: item.product.sku) %></strong>
```
Comment (elle) [inline app/views/spree/admin/orders/_stock_item.html.erb]: :+1:

File: app/views/spree/admin/orders/_stock_item.html.erb [hunk 7]
```diff
@@ -1,45 +1,45 @@
...
-      <% if item.part %>
-        ---
-      <% else %>
-        <%= item.line_item.single_money.to_html if item.line_item %>
+
+    <td class="item-price text-center"><%= item.line_item.single_money.to_html %></td>
+
+    <td class="item-qty-show text-center">
+      <% item.states.each do |state,count| %>
+        <%= count %> x <%= Spree.t(state).downcase %>
```
Comment (elle) [inline app/views/spree/admin/orders/_stock_item.html.erb]: Do we want to do the `downcase` in css?

File: spree_product_assembly.gemspec [hunk 8]
```diff
@@ -29,4 +29,5 @@ Gem::Specification.new do |s|
   s.add_development_dependency 'database_cleaner', '~> 1.4'
   s.add_development_dependency 'simplecov'
   s.add_development_dependency 'pg'
+  s.add_development_dependency 'launchy'
```
Comment (elle) [inline spree_product_assembly.gemspec]: Alphabetise??

File: app/views/spree/admin/orders/_assemblies.html.erb (hunk 2 above)
Comment (sikachu) [inline app/views/spree/admin/orders/_assemblies.html.erb]: Yeah ... it feels like that will make it not violate LoD. However, I'm not sure it's worth opening the class, risking method collision, and define those delegations though.

Comment (vinniefranco) [inline app/views/spree/admin/orders/_assemblies.html.erb]: Agreed

**Pull request merged**